http: Set TCP keepalive

Lazin opened the pull request:
Set keepalive on HTTP sockets.

<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [ ] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [x] v25.1.x
- [x] v24.3.x
- [x] v24.2.x

## Release Notes

<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->
### Bug Fixes

* Enable TCP keepalive for cloud storage connections.

Lazin committed:
net: Add set_keepalive_parameters method

to the base_transport. The method allows us to propagate the keepalive
parameters to the underlying TCP socket.

Signed-off-by: Evgeny Lazin <4lazin@gmail.com>

dotnwat reviewed (commented):
needs commit message. what is it fixing, why does it fix it, how was it found etc...

Lazin commented:
> needs commit message. what is it fixing, why does it fix it, how was it found etc...

updated

Lazin committed:
http: Set keepalive for http

The http client can potentially deadlock waiting for data from the
connection. If for some reason the connection was dropped and the RST
packet wasn't propagated the `recv_some` call in the http client will be
blocked for arbitrary amount of time (in some cases the middleware could
detect this and send the RST packet later).

If the keepalive is enabled we should send the proble after the idle
time (set to 5m by this commit). The defaults are chosen to be slightly
higher compared to our defaults for Kafka.

Signed-off-by: Evgeny Lazin <4lazin@gmail.com>

Lazin committed:
net: Add set_keepalive method to base_transport

The method enables keepalive timer on a socket.

Signed-off-by: Evgeny Lazin <4lazin@gmail.com>

StephanDollberg reviewed (dismissed):
Don't think there is any downside to enabling these.

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--d8f56fed-dce3-47d7-9fec-d16a374114c4-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/67112">build#67112</a></summary>


| test_class                | test_method      | test_arguments   | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                  |
|:--------------------------|:-----------------|:-----------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| CloudStorageChunkReadTest | test_read_chunks |                  | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67112#01975b44-c18c-4a16-b988-5b741f0fa046 | FLAKY         | 20/21    | upstream reliability is '97.6271186440678'. current run reliability is '95.23809523809523'. drift is 2.38902 and the allowed drift is set to 50. The test should PASS   |
| DisablingPartitionsTest   | test_disable     |                  | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67112#01975b44-c18d-49ab-8470-b11e1e51a4c2 | FLAKY         | 17/21    | upstream reliability is '92.88702928870293'. current run reliability is '80.95238095238095'. drift is 11.93465 and the allowed drift is set to 50. The test should PASS |
</details>
<!--/body-->

dotnwat commented on src/v/http/client.cc:
```diff
@@ -150,6 +155,14 @@ ss::future<client::request_response_t> client::make_request(
           }
+          // Set keepalive after connection is established and we have a
+          // connected socket.
+          set_keepalive_parameters(ss::net::tcp_keepalive_params{
+            .idle = tcp_keepalive_idle,
+            .interval = tcp_keepalive_interval,
+            .count = tcp_keepalive_probes,
+          });
+          set_keepalive(true);
```
should these be configurable, or this is good enough cc @StephanDollberg

StephanDollberg replied:
I mean configurable is always good but I don't really think it's needed. If anything you'd probably want to make the main toggle configurable

vbotbuildovich commented:
/backport v25.1.x

vbotbuildovich commented:
/backport v24.3.x

vbotbuildovich commented:
/backport v24.2.x
